File: backend/internal/application/word_service.go
package application

import "hotseat/backend/internal/domain"

// WordService exposes vocabulary queries and management to callers.
type WordService struct {
	words domain.WordRepository
}

func NewWordService(words domain.WordRepository) *WordService {
	return &WordService{words: words}
}

func (w *WordService) CountByDifficulty() (map[domain.Difficulty]int64, error) {
	return w.words.CountByDifficulty()
}

func (w *WordService) SetEnabled(id int64, enabled bool) error {
	return w.words.SetEnabled(id, enabled)
}

File: frontend/src/application/wordService.ts
import type { Category, Difficulty, Word, WordFilter } from '@/domain/word';
import type { WordRepository } from '@/infrastructure/repositories/wordRepository';

export class WordService {
  constructor(private readonly words: WordRepository) {}

  list(filter: WordFilter = {}): Word[] {
    return this.words.findAll(filter);
  }

  countsByDifficulty(): Record<Difficulty, number> {
    return this.words.countByDifficulty();
  }

  /** Cards per difficulty that have not appeared yet — the remaining pool. */
  remainingByDifficulty(): Record<Difficulty, number> {
    return this.words.countUnseenByDifficulty();
  }

  categories(): Category[] {
    return this.words.listCategories();
  }

  setEnabled(id: number, enabled: boolean): void {
    this.words.setEnabled(id, enabled);
  }
}
